File: scripts/qemu.py
"""QEMU embedded test commands."""

import argparse
import shutil
import subprocess
import sys

from shared.constants import QEMU_DIR
from shared.run import run

ARM_TARGETS = [
    ("thumbv7m-none-eabi", "arm,sqlite"),
    ("thumbv6m-none-eabi", "arm,sqlite"),
    ("thumbv8m.main-none-eabihf", "arm,sqlite"),
]

RISCV_TARGETS = [
    ("riscv32imc-unknown-none-elf", "riscv,sqlite"),
    ("riscv32imac-unknown-none-elf", "riscv,sqlite"),
]

ALL_TARGETS = ARM_TARGETS + RISCV_TARGETS


def _cargo_qemu(target: str, features: str, subcommand: str) -> list[str]:
    return [
        "cargo", subcommand,
        "--profile", "qemu",
        "--no-default-features",
        "--features", features,
        "--target", target,
        "--manifest-path", str(QEMU_DIR / "Cargo.toml"),
    ]


def _cargo_check(target: str, features: str) -> list[str]:
    return [
        "cargo", "check",
        "--no-default-features",
        "--features", features,
        "--target", target,
        "--manifest-path", str(QEMU_DIR / "Cargo.toml"),
    ]


def setup() -> int:
    if shutil.which("qemu-system-arm") and shutil.which("qemu-system-riscv32"):
        return 0
    if sys.platform == "darwin":
        return run(["brew", "install", "qemu"])
    elif sys.platform.startswith("linux"):
        return run(["sh", "-c", "sudo apt-get update && sudo apt-get install -y qemu-system"])
    print("[ERROR] Could not determine how to install QEMU on this platform.", file=sys.stderr)
    return 1


def build(target: str, features: str) -> int:
    return run(_cargo_qemu(target, features, "build"))


def build_arm() -> int:
    target, features = ARM_TARGETS[0]
    return build(target, features)


def build_riscv() -> int:
    target, features = RISCV_TARGETS[-1]
    return build(target, features)


def build_all() -> int:
    for target, features in ALL_TARGETS:
        rc = build(target, features)
        if rc != 0:
            return rc
    return 0


def run_arm() -> int:
    target, features = ARM_TARGETS[0]
    rc = build(target, features)
    if rc != 0:
        return rc
    return run([
        "qemu-system-arm", "-cpu", "cortex-m3", "-machine", "mps2-an385",
        "-nographic", "-semihosting-config", "enable=on,target=native",
        "-kernel", str(QEMU_DIR / "target" / target / "qemu" / "thumbv7m"),
    ])


def run_riscv() -> int:
    target, features = RISCV_TARGETS[-1]
    rc = build(target, features)
    if rc != 0:
        return rc
    return run([
        "qemu-system-riscv32", "-machine", "virt",
        "-nographic", "-semihosting-config", "enable=on,target=native",
        "-kernel", str(QEMU_DIR / "target" / target / "qemu" / "riscv32imac"),
    ])


def check() -> int:
    rc = 0
    for target, features in ALL_TARGETS:
        if run(_cargo_check(target, features)) != 0:
            rc = 1
    return rc


def clean() -> int:
    return run(["cargo", "clean"], cwd=QEMU_DIR)


def main() -> None:
    commands = {
        "setup": setup,
        "build_arm": build_arm,
        "build_riscv": build_riscv,
        "build_all": build_all,
        "run_arm": run_arm,
        "run_riscv": run_riscv,
        "check": check,
        "clean": clean,
    }

    parser = argparse.ArgumentParser(description="QEMU embedded test commands")
    parser.add_argument("command", nargs="?", default="check", choices=list(commands))
    args = parser.parse_args()
    sys.exit(commands[args.command]())


if __name__ == "__main__":
    main()

File: tools/xtask/src/qemu.rs
use anyhow::Context;

use crate::paths;
use crate::run;

pub const ARM_TARGETS: &[(&str, &str)] = &[
    ("thumbv7m-none-eabi", "arm"),
    ("thumbv6m-none-eabi", "arm,sqlite"),
    ("thumbv8m.main-none-eabihf", "arm,sqlite"),
];

pub const RISCV_TARGETS: &[(&str, &str)] = &[
    ("riscv32imc-unknown-none-elf", "riscv,sqlite"),
    ("riscv32imac-unknown-none-elf", "riscv,sqlite"),
];

const EMBEDDED_RUNNERS: &[(&str, &str, &[&str])] = &[
    (
        "thumbv7m-none-eabi",
        "arm",
        &[
            "qemu-system-arm",
            "-cpu",
            "cortex-m3",
            "-machine",
            "netduino2",
            "-nographic",
            "-semihosting-config",
            "enable=on,target=native",
            "-kernel",
        ],
    ),
    (
        "thumbv6m-none-eabi",
        "arm,sqlite",
        &[
            "qemu-system-arm",
            "-machine",
            "mps2-an385",
            "-nographic",
            "-semihosting-config",
            "enable=on,target=native",
            "-kernel",
        ],
    ),
    (
        "thumbv8m.main-none-eabihf",
        "arm,sqlite",
        &[
            "qemu-system-arm",
            "-cpu",
            "cortex-m33",
            "-machine",
            "mps2-an505",
            "-nographic",
            "-semihosting-config",
            "enable=on,target=native",
            "-kernel",
        ],
    ),
    (
        "riscv32imc-unknown-none-elf",
        "riscv,sqlite",
        &[
            "qemu-system-riscv32",
            "-M",
            "virt",
            "-bios",
            "none",
            "-nographic",
            "-semihosting-config",
            "enable=on,target=native",
            "-kernel",
        ],
    ),
    (
        "riscv32imac-unknown-none-elf",
        "riscv,sqlite",
        &[
            "qemu-system-riscv32",
            "-M",
            "virt",
            "-bios",
            "none",
            "-nographic",
            "-semihosting-config",
            "enable=on,target=native",
            "-kernel",
        ],
    ),
];

fn root() -> std::path::PathBuf {
    paths::repo_root()
}

fn runner_bin(target: &str) -> &'static str {
    match target {
        "thumbv7m-none-eabi" => "thumbv7m",
        "thumbv6m-none-eabi" => "thumbv6m",
        "thumbv8m.main-none-eabihf" => "thumbv8m",
        "riscv32imc-unknown-none-elf" => "riscv32imc",
        "riscv32imac-unknown-none-elf" => "riscv32imac",
        _ => unreachable!("no runner bin mapped for target {target}"),
    }
}

fn build_args(target: &str, features: &str) -> Vec<String> {
    vec![
        "build".into(),
        "--profile".into(),
        "qemu".into(),
        "--no-default-features".into(),
        "--features".into(),
        features.into(),
        "--target".into(),
        target.into(),
        "--bin".into(),
        runner_bin(target).into(),
        "--manifest-path".into(),
        paths::tests_dir().join("Cargo.toml").display().to_string(),
    ]
}

fn check_args(target: &str, features: &str) -> Vec<String> {
    vec![
        "check".into(),
        "--no-default-features".into(),
        "--features".into(),
        features.into(),
        "--target".into(),
        target.into(),
        "--bin".into(),
        runner_bin(target).into(),
        "--manifest-path".into(),
        paths::tests_dir().join("Cargo.toml").display().to_string(),
    ]
}

fn selftest_cargo(args: &[String]) -> anyhow::Result<()> {
    run::cargo(
        &root(),
        &args.iter().map(String::as_str).collect::<Vec<_>>(),
    )
}

fn qemu_bin_path(target: &str, bin: &str) -> anyhow::Result<std::path::PathBuf> {
    let path = paths::repo_root()
        .join("target")
        .join(target)
        .join("qemu")
        .join(bin);
    if !path.is_file() {
        anyhow::bail!(
            "{} does not exist; run `cargo xtask qemu build` first",
            path.display()
        );
    }
    Ok(path)
}

fn run_runner(target: &str, bin: &str, argv: &[&str]) -> anyhow::Result<()> {
    let el = qemu_bin_path(target, bin)?;
    let (program, args) = argv.split_first().expect("runner argv is never empty");
    let mut args = args.to_vec();
    let el_str = el.display().to_string();
    args.push(el_str.as_str());
    run::run(&root(), program, &args).with_context(|| format!("{bin} under QEMU"))
}

pub fn setup() -> anyhow::Result<()> {
    for bin in ["qemu-system-arm", "qemu-system-riscv32"] {
        if !run::which(bin) {
            anyhow::bail!(
                "{bin} not found. Install QEMU via your package manager \
                 (homebrew: `brew install qemu`; apt: `apt-get install qemu-system`)"
            );
        }
    }
    println!("qemu: system-arm + system-riscv32 present");
    Ok(())
}

pub fn build_arm() -> anyhow::Result<()> {
    let (target, features) = ARM_TARGETS[0];
    selftest_cargo(&build_args(target, features))
}

pub fn build_riscv() -> anyhow::Result<()> {
    let (target, features) = RISCV_TARGETS[1];
    selftest_cargo(&build_args(target, features))
}

pub fn build_all() -> anyhow::Result<()> {
    for (target, features) in ARM_TARGETS.iter().chain(RISCV_TARGETS) {
        selftest_cargo(&build_args(target, features))?;
    }
    Ok(())
}

pub fn check() -> anyhow::Result<()> {
    setup()?;
    for (target, features) in ARM_TARGETS.iter().chain(RISCV_TARGETS) {
        selftest_cargo(&check_args(target, features))?;
    }
    Ok(())
}

pub fn test_embedded() -> anyhow::Result<()> {
    setup()?;
    for (target, features, argv) in EMBEDDED_RUNNERS {
        selftest_cargo(&build_args(target, features))?;
        run_runner(target, runner_bin(target), argv)?;
    }
    Ok(())
}

pub fn run_arm() -> anyhow::Result<()> {
    setup()?;
    build_arm()?;
    let argv = &[
        "qemu-system-arm",
        "-cpu",
        "cortex-m3",
        "-machine",
        "mps2-an385",
        "-nographic",
        "-semihosting-config",
        "enable=on,target=native",
        "-kernel",
    ];
    run_runner("thumbv7m-none-eabi", "thumbv7m", argv)
}

pub fn run_riscv() -> anyhow::Result<()> {
    setup()?;
    build_riscv()?;
    let argv = &[
        "qemu-system-riscv32",
        "-M",
        "virt",
        "-nographic",
        "-semihosting-config",
        "enable=on,target=native",
        "-kernel",
    ];
    run_runner("riscv32imac-unknown-none-elf", "riscv32imac", argv)
}

pub fn clean() -> anyhow::Result<()> {
    selftest_cargo(&["clean".to_string()])
}
